Foundation/UserDefaults.swift

```diff
@@ -469,30 +349,44 @@ open class UserDefaults: NSObject {
     
     open func setPersistentDomain(_ domain: [String : Any], forName domainName: String) {
         if let defaults = UserDefaults(suiteName: domainName) {
+            var removedAny = false
             for key in defaults._dictionaryRepresentation(searchingOutsideOfSuite: false).keys {
                 defaults.removeObject(forKey: key)
+                removedAny = true
             }
             
+            var addedAny = false
             for (key, value) in domain {
                 defaults.set(value, forKey: key)
+                addedAny = true
+            }
+            
+            _ = defaults.synchronize()
+            
+            if removedAny || addedAny {
+                NotificationCenter.default.post(name: UserDefaults.didChangeNotification, object: self)
             }
         }
-        
-        NotificationCenter.default.post(name: UserDefaults.didChangeNotification, object: self)
     }
     
     open func removePersistentDomain(forName domainName: String) {
         if let defaults = UserDefaults(suiteName: domainName) {
+            var removedAny = false
             for key in defaults._dictionaryRepresentation(searchingOutsideOfSuite: false).keys {
                 defaults.removeObject(forKey: key)
+                removedAny = true
             }
             
-            NotificationCenter.default.post(name: UserDefaults.didChangeNotification, object: self)
+            _ = defaults.synchronize()
+            
+            if removedAny {
+                NotificationCenter.default.post(name: UserDefaults.didChangeNotification, object: self)
+            }
         }
     }
     
     open func synchronize() -> Bool {
-        return CFPreferencesAppSynchronize(kCFPreferencesCurrentApplication)
+        return CFPreferencesAppSynchronize(suite?._cfObject ?? kCFPreferencesCurrentApplication)
```

Comment:
Something I did in the ObjC implementation is add CF SPI for “synchronize everything” and call it here, since a ubiquitous pattern of misuse is synchronizing standard user defaults when you set something else.

Reply:
… ah. hm. I'm inclined to have a follow-up on this one.